wandb/history.py

```diff
@@ -0,0 +1,157 @@
+        self._add_callback = add_callback
+
+    def keys(self):
+        return list(self._keys)
+
+    def stream(self, name):
+        """stream can be used to record different time series:
+
+        run.history.stream("batch").add({"gradients": 1})
+        """
+        if self.stream_name != "default":
+            raise ValueError("Nested streams aren't supported")
+        if self._streams.get(name) == None:
+            self._streams[name] = History(self.fname, out_dir=self.out_dir,
+                                          add_callback=self._add_callback, stream_name=name)
+        return self._streams[name]
+
+    def column(self, key):
+        """Fetches a key from all rows that have it. Skips those that don't.
+        """
+        for row in self.rows:
+            if key in row:
+                yield row[key]
+
+    def add(self, row={}):
+        """Adds keys to history and writes the row.  If row isn't specified, will write
+        the current state of row.
+
+        run.history.row["duration"] = 1.0
+        run.history.add({"loss": 1})
+        => {"duration": 1.0, "loss": 1}
+
+        """
+        if not isinstance(row, collections.Mapping):
+            raise wandb.Error('history.add expects dict-like object')
+        self.row.update(row)
+        if not self.batched:
+            self._write()
+
+    @contextlib.contextmanager
+    def step(self, compute=True):
+        """Context manager to gradually build a history row, then commit it at the end.
+
+        To reduce the number of conditionals needed, code can check run.history.compute:
+
+        run.history.step(batch_idx % log_interval == 0):
+            run.history.add({"nice": "ok"})
+            if run.history.compute:
+                # Something expensive here
+        """
+        self.row = {}
+        self.batched = True
+        self.compute = compute
+        yield self
+        if compute:
+            self._write()
+
+    @property
+    def torch(self):
+        if self._torch is None:
+            self._torch = TorchHistory(self)
+        return self._torch
+
+    def _index(self, row):
+        """Internal row adding method that updates step, and keys"""
+        # TODO: store a downsampled representation in memory
+        self.rows.append(row)
+        self._keys.update(row.keys())
+        self._steps += 1
+
+    def _transform(self):
+        """Transforms media classes into the proper format before writing"""
+        for key, val in six.iteritems(self.row):
+            if type(val) in (list, tuple):
+                if len(val) > 0 and type(val[0]) == media.Image:
+                    self.row[key] = media.Image.transform(val, self.out_dir,
+                                                          "{}_{}.jpg".format(key, self.row["_step"]))
+
+    def _write(self):
+        if self.row:
+            self._lock.acquire()
+            try:
+                self.row['_runtime'] = round(time.time() - self._start_time, 2)
```

Comment:
Should we save wall-clock time while we're at it? Would make it easy to make charts that work across runs.

Also, I know we talked about the rounding before but now I'm feeling like, why drop information here. We can round in the UI.